joomla/joomla-cms
# [3.9] Pre Update Checker Error descriptions

Pull Request for Issue #17820 .

### Summary of Changes

When the user wants to update to a new Joomla version but has a PHP version installed or uses a database type (sqlsrv or sqlazure when updating to Joomla 4.0) that is not supported by the new version, new error messages are shown to the user. Previously, in all of these cases, the notice "We can't find a download URL" was shown. This hint is now only shown, when the user has a compatible php version installed and uses a compatible DB type. Therefore, the JUpdate class now always returns the minimum php version, even when the currently installed php version is not compatible.


### Testing Instructions

1. Create file `list.xml` at `/`
```xml
<extensionset>
    <extension element="joomla" type="file" version="4.0.0" targetplatformversion="3.9" detailsurl="http://localhost/extension_sts.xml" />
</extensionset>
```
 
2. Create file `extension_sts.xml` at `/`
```xml
<updates>
    <update>
        <name>Joomla! 3.7</name>
        <description>Joomla! 3.7 CMS</description>
        <element>joomla</element>
        <type>file</type>
        <version>4.0.0</version>
        <infourl title="Joomla!">https://www.joomla.org/announcements/release-news/5711-joomla-3-7-5-release.html</infourl>
        <downloads>
            <downloadurl type="full" format="zip">https://downloads.joomla.org/cms/joomla3/3-7-5/Joomla_3.7.5-Stable-Update_Package.zip</downloadurl>
        </downloads>
        <tags>
            <tag>stable</tag>
        </tags>
        <maintainer>Joomla! PLT</maintainer>
        <maintainerurl>https://www.joomla.org</maintainerurl>
        <section>STS</section>
        <targetplatform name="joomla" version="3.[23456789]" />
        <php_minimum>7.0</php_minimum>
    </update>
</updates>
```

3. Components -> Joomla! Update -> Options
 
Custom URL: `localhost/list.xml`

4. Save & Close
 
There should be a fake update to Joomla 4.0 and you should see the pre update checker from #17742

5. When changing the version in <php_minimum>-tag to 8 in the file `extension_sts.xml` at `/` you should see a hint that your current php version is not supported.

6. Add your current database type (e.g. `mysqli`) to `$unsupportedDatabaseTypes` in function `isDatabaseTypeSupported()` in file `./administrator/components/com_joomlaupdate/models/default.php`.

7. You should now see two error descriptions, one for the not supported php version and one for the used db type being not compatible.


![updateerrors](https://user-images.githubusercontent.com/24227181/30005441-0141c252-90e3-11e7-9df0-fc07d67ebd1f.png)

### Expected result

You see different error descriptions depending on the error preventing the update to be installed. (Wrong PHP Version, wrong DB type, problem with update server).

### Actual result

You only see one error description, no matter if the download fails due to a problem with the Joomla Update Server or due to your server not meeting the minimum update requirements.

### Documentation Changes Required

None.
